From acb9a4875bd328f823f77ce842495cd88c89ee96 Mon Sep 17 00:00:00 2001
From: Robert Beckett <bob.beckett@collabora.com>
Date: Thu, 12 Mar 2026 17:27:25 +0000
Subject: [PATCH 040/188] PM: recover devices that refused to thaw on late
 abandonded hibernation

When a device resfuses to thaw for hibernation, if the hibernation
is abandoned late after device handling has finished, the devices
will be left on the pm_list still suspended.
As the hibernation is abandoned, there is no restore path and the
existing thaw infrastructure thinks that the device is already
thawed.

Search for suspended devices and issue a recover resume to
let them handle the abandoned hibernation state.

Signed-off-by: Robert Beckett <bob.beckett@collabora.com>
---
 drivers/base/power/main.c | 46 +++++++++++++++++++++++++++++++++++++++
 include/linux/pm.h        |  1 +
 kernel/power/hibernate.c  | 15 ++++++++-----
 3 files changed, 57 insertions(+), 5 deletions(-)

diff --git a/drivers/base/power/main.c b/drivers/base/power/main.c
index 510aa4cb6..91e90c67c 100644
--- a/drivers/base/power/main.c
+++ b/drivers/base/power/main.c
@@ -1265,6 +1265,52 @@ void dpm_resume(pm_message_t state)
 	trace_suspend_resume(TPS("dpm_resume"), state.event, false);
 }
 
+void dpm_resume_suspended_devices(pm_message_t state)
+{
+	struct device *dev;
+	ktime_t starttime = ktime_get();
+
+	trace_suspend_resume(TPS("dpm_resume_suspended_devices"), state.event, true);
+
+	pm_transition = state;
+	async_error = 0;
+
+	mutex_lock(&dpm_list_mtx);
+
+	/*
+	 * Start processing "async" root devices upfront so they don't wait for
+	 * the "sync" devices they don't depend on.
+	 */
+	list_for_each_entry(dev, &dpm_list, power.entry) {
+		if (!dev->power.is_suspended)
+			continue;
+
+		dpm_clear_async_state(dev);
+		if (dpm_root_device(dev))
+			dpm_async_with_cleanup(dev, async_resume);
+
+		if (!dpm_async_fn(dev, async_resume)) {
+			get_device(dev);
+
+			mutex_unlock(&dpm_list_mtx);
+
+			device_resume(dev, state, false);
+
+			put_device(dev);
+
+			mutex_lock(&dpm_list_mtx);
+		}
+	}
+
+	mutex_unlock(&dpm_list_mtx);
+	async_synchronize_full();
+	dpm_show_time(starttime, state, 0, NULL);
+	if (READ_ONCE(async_error))
+		dpm_save_failed_step(SUSPEND_RESUME);
+
+	trace_suspend_resume(TPS("dpm_resume_suspended_devices"), state.event, false);
+}
+
 /**
  * device_complete - Complete a PM transition for given device.
  * @dev: Device to handle.
diff --git a/include/linux/pm.h b/include/linux/pm.h
index 06d3029af..d61a50e95 100644
--- a/include/linux/pm.h
+++ b/include/linux/pm.h
@@ -822,6 +822,7 @@ extern void dpm_resume_end(pm_message_t state);
 extern void dpm_resume_noirq(pm_message_t state);
 extern void dpm_resume_early(pm_message_t state);
 extern void dpm_resume(pm_message_t state);
+extern void dpm_resume_suspended_devices(pm_message_t state);
 extern void dpm_complete(pm_message_t state);
 
 extern void device_pm_unlock(void);
diff --git a/kernel/power/hibernate.c b/kernel/power/hibernate.c
index 1a7da16dc..3911d76a4 100644
--- a/kernel/power/hibernate.c
+++ b/kernel/power/hibernate.c
@@ -672,13 +672,14 @@ int hibernation_platform_enter(void)
  * state corresponding to hibernation, or try to power it off or reboot,
  * depending on the value of hibernation_mode.
  */
-static void power_down(void)
+static int power_down(void)
 {
-	int error = -EAGAIN;
+	int error = 0;
 
 	if (pm_wakeup_pending()) {
 		pm_wakeup_clear(0);
 		pr_info("Wakeup event detected after image write, aborting power down.\n");
+		error = -EAGAIN;
 		goto exit;
 	}
 
@@ -724,9 +725,9 @@ static void power_down(void)
 
 exit:
 	/* Restore swap signature. */
-	error = swsusp_unmark();
-	if (error)
+	if (swsusp_unmark())
 		pr_err("Swap will be unusable! Try swapon -a.\n");
+	return error;
 }
 
 static int load_image_and_restore(void)
@@ -848,7 +849,11 @@ int hibernate(void)
 			if (hibernation_mode == HIBERNATION_TEST_RESUME)
 				snapshot_test = true;
 			else
-				power_down();
+				error = power_down();
+		}
+		if (error) {
+			/* recover any devices that refused to thaw */
+			dpm_resume_suspended_devices(PMSG_RECOVER);
 		}
 		in_suspend = 0;
 		pm_restore_gfp_mask();
-- 
2.55.0

